ironfish/src/rpc/stream.ts

```diff
@@ -1,32 +1,66 @@
 /* This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
-import { PromiseResolve } from '../utils'
+import { PromiseReject, PromiseResolve } from '../utils'
 
+/*
+ * A Stream takes a stream of data and transforms it into an iterable
+ * On one end the writer of the data is calling stream.write(data)
+ * and on the other end the consumer of the data is calling stream.next()
+ *
+ * If stream.next() is called before data is written then the request will be queued
+ * up in the `waiting` list by saving functions to resolve and reject the request
+ *
+ * stream.next()      -> [(res, rej)]               a resolve and reject for the data is queued in `waiting`
+ * stream.next()      -> [(res, rej), (res, rej)]
+ * stream.write(data) -> [(res, rej)]               the first Promise in waiting is resolved with the data
+ * stream.write(data) -> []                         the final Promise in waiting is resolved with the data
+ *
+ * If stream.write() is called before the reader requests with stream.next() then the
+ * data will be queued up in the `buffer` list
+ *
+ * stream.write(1) -> [1]                            data is queued up in the buffer
+ * stream.write(2) -> [1, 2]
+ * stream.write(3) -> [1, 2, 3]
+ * stream.next()   -> [2, 3]                         data is consumed from the buffer and resolved to next()
+ * stream.next()   -> [3]
+ *
+ * Because Stream implements AsyncIterable it can also be used in for...in loops
+ */
 export class Stream<T> implements AsyncIterable<T> {
   buffer: T[] = []
-  waiting: PromiseResolve<IteratorResult<T>>[] = []
+  waiting: [PromiseResolve<IteratorResult<T>>, PromiseReject][] = []
+
+  // Stream can either be open or closed
+  // It can be closed because of an error or just closed normally
```

Comment:
I think this whole comment doesn't seem useful to me personally

```suggestion
```

Reply:
Ok I can remove this comment